src/rust/lib/graph-editor/src/component/visualization.rs

```diff
@@ -0,0 +1,332 @@
+            def update_visualization    = source::<Option<Visualization>> ();
+            def update_data       = source::<Data>                  ();
+        };
+        let network = visualization_events;
+        Self {network,show,hide,update_visualization,toggle_visibility,update_data}
+    }
+}
+
+
+
+// ================================
+// === Visualizations Container ===
+// ================================
+
+/// Container that wraps a `Visualisation` for rendering and interaction in the GUI.
+#[derive(Clone,CloneRef,Debug)]
+#[allow(missing_docs)]
+pub struct Container {
+    pub data : Rc<ContainerData>
+}
+
+/// Weak version of `Container`.
+#[derive(Clone,CloneRef,Debug)]
+pub struct WeakContainer {
+    data : Weak<ContainerData>
+}
+
+/// Internal data of a `Container`.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct ContainerData {
+    pub logger : Logger,
+    pub events : Events,
+
+    node          : display::object::Instance,
+    size          : Cell<Vector2<f32>>,
+    position      : Cell<Vector3<f32>>,
+    visible       : Cell<bool>,
+    visualization : RefCell<Option<Visualization>>,
+}
+
+impl Container {
+    /// Constructor.
+    pub fn new() -> Self {
+        let logger   = Logger::new("visualization");
+        let events   = Events::default();
+        let content  = RefCell::new(None);
+        let size     = Cell::new(Vector2::new(100.0, 100.0));
+        let position = Cell::new(Vector3::new(  0.0,-110.0, 0.0));
+        let visible  = Cell::new(true);
+        let node     = display::object::Instance::new(&logger);
+
+        let data     = ContainerData {logger,events, visualization: content,size,position,visible,node};
+        let data     = Rc::new(data);
+        Self {data} . init_frp()
+    }
+
+    /// Update the content properties with the values from the `ContainerData`.
+    ///
+    /// Needs to called when a visualisation has been set.
+    fn update_visualisation_properties(&self) {
+        let size       = self.data.size.get();
+        let position   = self.data.position.get();
+
+        if let Some(vis) = self.data.visualization.borrow().as_ref() {
+            vis.content.set_size(size);
+            vis.content.set_position(position);
+        };
+    }
+
+    /// Set the visualization content.
+    pub fn set_visualisation(&self, visualization: Visualization) {
+        self.display_object().add_child(visualization.display_object());
+        self.data.visualization.replace(Some(visualization));
+        self.update_visualisation_properties();
+    }
+
+    fn init_frp(self) -> Self {
+        let network = &self.data.events.network;
+
+        frp::extend! { network
+            let weak_vis = self.downgrade();
+            def _f_show = self.data.events.show.map(move |_| {
+               if let Some(vis) = weak_vis.upgrade() {
+                    vis.set_visibility(true)
+               }
+            });
```

Comment:
Here is a little bit ugly pattern - you need to do all this fancy stuff with weak and strong refs only because the lambdas in FRP are referencing the same object that owns the `frp::Network` (which owns the lambdas). Maybe you are able to somehow split it? So there would be struct owning `frp::Network` and another rc struct, and the lambdas could access only the another struct? In such case you would not need anymore to do all this upgrade / downgrade ping-pong. Maybe its impossible (I don't know yet, but in case it is possible, please do it).

Reply:
I've refactored this, so there are two structs: one for the data/logic (`ContainerData`) and one for the FRP (`ContainerFrp`), which are both held through `Rc`s by the `Container`/`WeakContainer`. If I'm seeing this right, that means there should be no cycle between the two, as the reference structure looks like this:
```text
                     `Container`/`WeakContainer`
                       /                      \
                      /                        \
                     V                          V
         `ContainerFrp`        ------>         `ContainerData`

```